modules/gltf/src/lib/extensions/KHR_texture_transform.ts

```diff
@@ -0,0 +1,145 @@
+import type {GLTFLoaderOptions} from '../../gltf-loader';
+import {BYTES, COMPONENTS} from '../api/post-process-gltf';
+import {getAccessorArrayTypeAndLength} from '../gltf-utils/gltf-utils';
+
+/** Extension name */
+const EXT_MESHOPT_TRANSFORM = 'KHR_texture_transform';
+
+export const name = EXT_MESHOPT_TRANSFORM;
+
+const scratchVector = new Vector3();
+const scratchRotationMatrix = new Matrix3();
+const scratchScaleMatrix = new Matrix3();
+
+export async function decode(
+  gltfData: {
+    json: GLTF;
+    buffers: {arrayBuffer: ArrayBuffer; byteOffset: number; byteLength: number}[];
+  },
+  options: GLTFLoaderOptions
+) {
+  const materials = gltfData.json.materials || [];
+  for (let i = 0; i < materials.length; i++) {
+    const material = materials[i];
+    const extensionData = getExtensionData(material);
+    if (extensionData) {
+      const matrix = makeTransformationMatrix(extensionData);
+      transformPrimitives(gltfData, i, matrix);
+    }
+  }
+}
+
+function getExtensionData(material: GLTFMaterial): any {
+  const baseColorTexture = material.pbrMetallicRoughness?.baseColorTexture;
+  if (baseColorTexture) {
+    return baseColorTexture.extensions?.[EXT_MESHOPT_TRANSFORM];
+  }
+  const emisiveTexture = material.emissiveTexture;
+  if (emisiveTexture) {
+    return emisiveTexture.extensions?.[EXT_MESHOPT_TRANSFORM];
+  }
+  const normalTexture = material.normalTexture;
+  if (normalTexture) {
+    return normalTexture.extensions?.[EXT_MESHOPT_TRANSFORM];
+  }
+  const occlusionTexture = material.occlusionTexture;
+  if (occlusionTexture) {
+    return occlusionTexture.extensions?.[EXT_MESHOPT_TRANSFORM];
+  }
+  const metallicRoughnessTexture = material.pbrMetallicRoughness?.metallicRoughnessTexture;
+  if (metallicRoughnessTexture) {
+    return metallicRoughnessTexture.extensions?.[EXT_MESHOPT_TRANSFORM];
+  }
+  return null;
+}
+
+function transformPrimitives(
+  gltfData: {
+    json: GLTF;
+    buffers: {arrayBuffer: ArrayBuffer; byteOffset: number; byteLength: number}[];
+  },
+  materialIndex: number,
+  transformationMatrix: Matrix3
+) {
+  const meshes = gltfData.json.meshes || [];
+  for (const mesh of meshes) {
+    for (const primitive of mesh.primitives) {
+      const material = primitive.material;
+      if (Number.isFinite(material) && materialIndex === material) {
+        transformPrimitive(gltfData, primitive, transformationMatrix);
+      }
+    }
+  }
+}
+
+function transformPrimitive(
+  gltfData: {
+    json: GLTF;
+    buffers: {arrayBuffer: ArrayBuffer; byteOffset: number; byteLength: number}[];
+  },
+  primitive: GLTFMeshPrimitive,
+  transformationMatrix: Matrix3
+) {
+  const texCoordAccessor = primitive.attributes.TEXCOORD_0;
+  if (Number.isFinite(texCoordAccessor)) {
+    const accessor = gltfData.json.accessors?.[texCoordAccessor];
+    if (accessor && accessor.bufferView) {
+      const bufferView = gltfData.json.bufferViews?.[accessor.bufferView];
```

Comment:
This code is very compact and hard to read. Maybe add one or two well named helper functions, and/or some comments.

Reply:
Done